### PR TITLE
cpu/sam0_common: implement periph_gpio_ll and periph_gpio_ll_irq

### PR DESCRIPTION
**Pull request by maribu**
### Contribution description

This implements `periph_gpio_ll` and `periph_gpio_ll_irq` for SAM0 based MCUs.

### Testing procedure

```
make BOARD=samr21-xpro flash test-with-config -C tests/periph/gpio_ll 
make: Entering directory '/home/maribu/Repos/software/RIOT/gpio_ll_sam0/tests/periph/gpio_ll'
Building application "tests_gpio_ll" for "samr21-xpro" with MCU "samd21".

"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/pkg/cmsis/ 
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/boards/common/init
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/boards/samr21-xpro
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/core
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/core/lib
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/samd21
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/cortexm_common
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/cortexm_common/periph
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/sam0_common
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/sam0_common/periph
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/samd21/periph
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/samd21/vectors
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/drivers
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/drivers/periph_common
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/auto_init
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/div
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/frac
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/isrpipe
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/libc
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/malloc_thread_safe
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/newlib_syscalls_default
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/pm_layered
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/preprocessor
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/stdio_uart
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/test_utils/interactive_sync
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/test_utils/print_stack_usage
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/tsrb
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/ztimer
   text	  data	   bss	   dec	   hex	filename
  21788	   176	  2708	 24672	  6060	/home/maribu/Repos/software/RIOT/gpio_ll_sam0/tests/periph/gpio_ll/bin/samr21-xpro/tests_gpio_ll.elf
/home/maribu/Repos/software/RIOT/gpio_ll_sam0/dist/tools/edbg/edbg.sh flash /home/maribu/Repos/software/RIOT/gpio_ll_sam0/tests/periph/gpio_ll/bin/samr21-xpro/tests_gpio_ll.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009225 03.25.01B6 (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification...
at address 0x40 expected 0xfd, read 0xf1
Error: verification failed
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009225 03.25.01B6 (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Programming...... done.
Verification...... done.
Done flashing
r
/home/maribu/Repos/software/RIOT/gpio_ll_sam0/dist/tools/pyterm/pyterm -p "/dev/ttyACM1" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Connect to serial port /dev/ttyACM1
Welcome to pyterm!
Type '/exit' to exit.
READY
s
START
main(): This is RIOT! (Version: 2024.01-devel-411-g5552b4-cpu/sam0_common/periph_gpio_ll)
Test / Hardware Details:
========================
Cabling:
(INPUT -- OUTPUT)
  P1.2 (PB2) -- P0.14 (PA14)
  P1.23 (PB23) -- P0.8 (PA8)
Number of pull resistor values supported: 1
Number of drive strengths supported: 2
Number of slew rates supported: 1
Valid GPIO ports:
- PORT 0 (PORT A)
- PORT 1 (PORT B)
- PORT 2 (PORT C)

Testing gpio_port_pack_addr()
=============================

All OK

Testing gpip_ng_init()
======================

Testing is_gpio_port_num_valid() is true for PORT_OUT and PORT_IN:

Testing input configurations for PIN_IN_0:
Support for input with pull up: yes
state: in, pull: up, schmitt trigger: on, value: on
Support for input with pull down: yes
state: in, pull: down, schmitt trigger: on, value: off
Support for input with pull to bus level: no
Support for floating input (no pull resistors): yes
state: in, pull: none, schmitt trigger: on, value: off

Testing output configurations for PIN_OUT_0:
Support for output (push-pull) with initial value of LOW: yes
state: out-pp, drive: weak, value: off
Output is indeed LOW: yes
state: out-pp, drive: weak, value: on
Output can be pushed HIGH: yes
Support for output (push-pull) with initial value of HIGH: yes
state: out-pp, drive: weak, value: on
Output is indeed HIGH: yes
Support for output (open drain with pull up) with initial value of LOW: no
Support for output (open drain with pull up) with initial value of HIGH: no
Support for output (open drain) with initial value of LOW: no
Support for output (open drain) with initial value of HIGH: no
Support for output (open source) with initial value of LOW: no
Support for output (open source) with initial value of HIGH: no
Support for output (open source with pull up) with initial value of HIGH: no
Support for output (open source with pull up) with initial value of LOW: no
Support for disconnecting GPIO: yes
state: off, drive: weak, pull: none, schmitt trigger: off, value: off
Output can indeed be pulled LOW: yes
Output can indeed be pulled HIGH: yes

Testing Reading/Writing GPIO Ports
==================================

testing initial value of 0 after init
...OK
testing setting both outputs_optional simultaneously
...OK
testing clearing both outputs_optional simultaneously
...OK
testing toggling first output (0 --> 1)
...OK
testing toggling first output (1 --> 0)
...OK
testing toggling second output (0 --> 1)
...OK
testing toggling second output (1 --> 0)
...OK
testing setting first output and clearing second with write
...OK
testing setting second output and clearing first with write
...OK
All input/output operations worked as expected

Testing External IRQs
=====================

Testing rising edge on PIN_IN_0
... OK
Testing falling edge on PIN_IN_0
... OK
Testing both edges on PIN_IN_0
... OK
Testing masking of IRQs (still both edges on PIN_IN_0)
... OK
Testing level-triggered on HIGH on PIN_IN_0 (when input is LOW when setting up IRQ)
... OK
Testing level-triggered on HIGH on PIN_IN_0 (when input is HIGH when setting up IRQ)
... OK
Testing level-triggered on LOW on PIN_IN_0 (when input is HIGH when setting up IRQ)
... OK
Testing level-triggered on LOW on PIN_IN_0 (when input is LOW when setting up IRQ)
... OK


TEST SUCCEEDED
```

### Issues/PRs references

- [x] Depends on and includes https://github.com/RIOT-OS/RIOT/pull/20236

### CODE REVIEW

**Comment by maribu** [comment]
Rebased on `master`, test still succeeds:

```
git:(cpu/sam0_common/periph_gpio_ll) ~/Repos/software/RIOT/gpio_ll_sam0 ➜ make BOARD=samr21-xpro -C tests/periph/gpio_ll flash test-with-config 
make: Entering directory '/home/maribu/Repos/software/RIOT/gpio_ll_sam0/tests/periph/gpio_ll'
Building application "tests_gpio_ll" for "samr21-xpro" with MCU "samd21".

"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/pkg/cmsis/ 
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/boards/common/init
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/boards/samr21-xpro
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/core
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/core/lib
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/samd21
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/cortexm_common
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/cortexm_common/periph
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/sam0_common
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/sam0_common/periph
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/samd21/periph
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/cpu/samd21/vectors
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/drivers
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/drivers/periph_common
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/auto_init
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/div
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/frac
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/isrpipe
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/libc
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/malloc_thread_safe
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/newlib_syscalls_default
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/pm_layered
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/preprocessor
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/stdio_uart
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/test_utils/interactive_sync
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/test_utils/print_stack_usage
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/tsrb
"make" -C /home/maribu/Repos/software/RIOT/gpio_ll_sam0/sys/ztimer
   text	  data	   bss	   dec	   hex	filename
  21204	   176	  2708	 24088	  5e18	/home/maribu/Repos/software/RIOT/gpio_ll_sam0/tests/periph/gpio_ll/bin/samr21-xpro/tests_gpio_ll.elf
/home/maribu/Repos/software/RIOT/gpio_ll_sam0/dist/tools/edbg/edbg.sh flash /home/maribu/Repos/software/RIOT/gpio_ll_sam0/tests/periph/gpio_ll/bin/samr21-xpro/tests_gpio_ll.bin
### Flashing Target ###
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009225 03.25.01B6 (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Verification...
at address 0x4 expected 0x6d, read 0x5d
Error: verification failed
Debugger: ATMEL EDBG CMSIS-DAP ATML2127031800009225 03.25.01B6 (S)
Clock frequency: 16.0 MHz
Target: SAM R21G18 (Rev C)
Programming...... done.
Verification...... done.
Done flashing
r
/home/maribu/Repos/software/RIOT/gpio_ll_sam0/dist/tools/pyterm/pyterm -p "/dev/ttyACM0" -b "115200" --no-reconnect --noprefix --no-repeat-command-on-empty-line 
Connect to serial port /dev/ttyACM0
Welcome to pyterm!
Type '/exit' to exit.
Y
 ��%���������х������������ɥ�сJсJ́�����READY
s
START
main(): This is RIOT! (Version: 2024.01-devel-659-g67b381-cpu/sam0_common/periph_gpio_ll)
Test / Hardware Details:
========================
Cabling:
(INPUT -- OUTPUT)
  P1.2 (PB2) -- P0.14 (PA14)
  P1.23 (PB23) -- P0.8 (PA8)
Number of pull resistor values supported: 1
Number of drive strengths supported: 2
Number of slew rates supported: 1
Valid GPIO ports:
- PORT 0 (PORT A)
- PORT 1 (PORT B)
- PORT 2 (PORT C)

Testing gpio_port_pack_addr()
=============================

All OK

Testing gpip_ng_init()
======================

Testing is_gpio_port_num_valid() is true for PORT_OUT and PORT_IN:

Testing input configurations for PIN_IN_0:
Support for input with pull up: yes
state: in, pull: up, value: on, drive: weak
Support for input with pull down: yes
state: in, pull: down, value: off, drive: weak
Support for input with pull to bus level: no
Support for floating input (no pull resistors): yes
state: in, pull: none, value: off, drive: weak

Testing output configurations for PIN_OUT_0:
Support for output (push-pull) with initial value of LOW: yes
state: out-pp, value: off, drive: weak
Output is indeed LOW: yes
state: out-pp, value: on, drive: weak
Output can be pushed HIGH: yes
Support for output (push-pull) with initial value of HIGH: yes
state: out-pp, value: on, drive: weak
Output is indeed HIGH: yes
Support for output (open drain with pull up) with initial value of LOW: no
Support for output (open drain with pull up) with initial value of HIGH: no
Support for output (open drain) with initial value of LOW: no
Support for output (open drain) with initial value of HIGH: no
Support for output (open source) with initial value of LOW: no
Support for output (open source) with initial value of HIGH: no
Support for output (open source with pull up) with initial value of HIGH: no
Support for output (open source with pull up) with initial value of LOW: no
Support for disconnecting GPIO: yes
state: off, pull: none, value: off, drive: weak
Output can indeed be pulled LOW: yes
Output can indeed be pulled HIGH: yes

Testing Reading/Writing GPIO Ports
==================================

testing initial value of 0 after init
...OK
testing setting both outputs_optional simultaneously
...OK
testing clearing both outputs_optional simultaneously
...OK
testing toggling first output (0 --> 1)
...OK
testing toggling first output (1 --> 0)
...OK
testing toggling second output (0 --> 1)
...OK
testing toggling second output (1 --> 0)
...OK
testing setting first output and clearing second with write
...OK
testing setting second output and clearing first with write
...OK
All input/output operations worked as expected

Testing External IRQs
=====================

Testing rising edge on PIN_IN_0
... OK
Testing falling edge on PIN_IN_0
... OK
Testing both edges on PIN_IN_0
... OK
Testing masking of IRQs (still both edges on PIN_IN_0)
... OK
Testing level-triggered on HIGH on PIN_IN_0 (when input is LOW when setting up IRQ)
... OK
Testing level-triggered on HIGH on PIN_IN_0 (when input is HIGH when setting up IRQ)
... OK
Testing level-triggered on LOW on PIN_IN_0 (when input is HIGH when setting up IRQ)
... OK
Testing level-triggered on LOW on PIN_IN_0 (when input is LOW when setting up IRQ)
... OK


TEST SUCCEEDED

make: Leaving directory '/home/maribu/Repos/software/RIOT/gpio_ll_sam0/tests/periph/gpio_ll'
```

**Comment by benpicco** [comment]
I tried this on `saml10-xpro` with 

```
PORT_IN=0
PORT_OUT=0
PIN_IN_0=2
PIN_OUT_0=3
PIN_IN_1=21
PIN_OUT_1=10
```

and I get

```
024-01-22 12:55:00,579 # main(): This is RIOT! (Version: 2024.01-devel-659-g67b381-HEAD)
2024-01-22 12:55:00,581 # Test / Hardware Details:
2024-01-22 12:55:00,584 # ========================
2024-01-22 12:55:00,584 # Cabling:
2024-01-22 12:55:00,586 # (INPUT -- OUTPUT)
2024-01-22 12:55:00,589 #   P0.2 (PA2) -- P0.3 (PA3)
2024-01-22 12:55:00,592 #   P0.21 (PA21) -- P0.10 (PA10)
2024-01-22 12:55:00,596 # Number of pull resistor values supported: 1
2024-01-22 12:55:00,600 # Number of drive strengths supported: 2
2024-01-22 12:55:00,603 # Number of slew rates supported: 1
2024-01-22 12:55:00,604 # Valid GPIO ports:
2024-01-22 12:55:00,606 # - PORT 0 (PORT A)
2024-01-22 12:55:00,606 # 
2024-01-22 12:55:00,609 # Testing gpio_port_pack_addr()
2024-01-22 12:55:00,612 # =============================
2024-01-22 12:55:00,612 # 
2024-01-22 12:55:00,612 # All OK
2024-01-22 12:55:00,613 # 
2024-01-22 12:55:00,615 # Testing gpip_ng_init()
2024-01-22 12:55:00,617 # ======================
2024-01-22 12:55:00,617 # 
2024-01-22 12:55:00,623 # Testing is_gpio_port_num_valid() is true for PORT_OUT and PORT_IN:
2024-01-22 12:55:00,623 # 
2024-01-22 12:55:00,627 # Testing input configurations for PIN_IN_0:
2024-01-22 12:55:00,630 # Support for input with pull up: yes
2024-01-22 12:55:00,634 # state: in, pull: up, value: on, drive: weak
2024-01-22 12:55:00,637 # Support for input with pull down: yes
2024-01-22 12:55:00,642 # state: in, pull: down, value: off, drive: weak
2024-01-22 12:55:00,646 # Support for input with pull to bus level: no
2024-01-22 12:55:00,650 # Support for floating input (no pull resistors): yes
2024-01-22 12:55:00,655 # state: in, pull: none, value: off, drive: weak
2024-01-22 12:55:00,655 # 
2024-01-22 12:55:00,659 # Testing output configurations for PIN_OUT_0:
2024-01-22 12:55:00,664 # Support for output (push-pull) with initial value of LOW: yes
2024-01-22 12:55:00,668 # state: out-pp, value: off, drive: weak
2024-01-22 12:55:00,671 # Output is indeed LOW: yes
2024-01-22 12:55:00,675 # state: out-pp, value: on, drive: weak
2024-01-22 12:55:00,679 # Output can be pushed HIGH: yes
2024-01-22 12:55:00,684 # Support for output (push-pull) with initial value of HIGH: yes
2024-01-22 12:55:00,688 # state: out-pp, value: on, drive: weak
2024-01-22 12:55:00,691 # Output is indeed HIGH: yes
2024-01-22 12:55:00,698 # Support for output (open drain with pull up) with initial value of LOW: no
2024-01-22 12:55:00,705 # Support for output (open drain with pull up) with initial value of HIGH: no
2024-01-22 12:55:00,711 # Support for output (open drain) with initial value of LOW: no
2024-01-22 12:55:00,716 # Support for output (open drain) with initial value of HIGH: no
2024-01-22 12:55:00,722 # Support for output (open source) with initial value of LOW: no
2024-01-22 12:55:00,728 # Support for output (open source) with initial value of HIGH: no
2024-01-22 12:55:00,735 # Support for output (open source with pull up) with initial value of HIGH: no
2024-01-22 12:55:00,742 # Support for output (open source with pull up) with initial value of LOW: no
2024-01-22 12:55:00,745 # Support for disconnecting GPIO: yes
2024-01-22 12:55:00,750 # state: off, pull: none, value: off, drive: weak
2024-01-22 12:55:00,754 # Output can indeed be pulled LOW: yes
2024-01-22 12:55:00,759 # Output can indeed be pulled HIGH: yes
2024-01-22 12:55:00,759 # 
2024-01-22 12:55:00,762 # Testing Reading/Writing GPIO Ports
2024-01-22 12:55:00,765 # ==================================
2024-01-22 12:55:00,765 # 
2024-01-22 12:55:00,768 # testing initial value of 0 after init
2024-01-22 12:55:00,770 # ...OK
2024-01-22 12:55:00,775 # testing setting both outputs_optional simultaneously
2024-01-22 12:55:00,778 # expect failed at line 526
```

**Comment by maribu** [comment]
PA21 seems to be connected to `EDBG_ADC0`, PA10 to `DGI_GPIO2`, and PA3 to `EDBG_CDC_USART_RX`.

Could you try it again with other GPIOs?

**Comment by benpicco** [comment]
Ah indeed, works with PA04/05, PA14/15

File: cpu/sam0_common/include/gpio_ll_arch.h [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+    return NULL;
+}
+
+static inline bool is_gpio_port_num_valid(uint_fast8_t num)
+{
+    return (num < ARRAY_SIZE(GPIO_IOBUS_BASE->Group));
+}
+
+
```
Comment (benpicco) [inline cpu/sam0_common/include/gpio_ll_arch.h:144]:
```suggestion
```

File: cpu/sam0_common/include/gpio_ll_arch.h [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ * @{
+ *
+ * @file
+ * @brief           CPU specific part of the Peripheral GPIO Low-Level API
+ *
+ * @author          Troels Hoffmeyer <troels.d.hoffmeyer@gmail.com>
+ * @author          Thomas Eichinger <thomas.eichinger@fu-berlin.de>
+ * @author          Kaspar Schleiser <kaspar@schleiser.de>
+ * @author          Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author          Juergen Fitschen <me@jue.yt>
```
Comment (benpicco) [inline cpu/sam0_common/include/gpio_ll_arch.h:22]: Not sure how many of those people are even aware of the GPIO LL API

**Comment by benpicco** [comment]
I get a different failure on `same54-xpro` though with

```
PORT_IN=1
PORT_OUT=3
PIN_IN_0=14
PIN_OUT_0=0
PIN_IN_1=15
PIN_OUT_1=9
```

```
2024-01-22 14:21:11,080 # main(): This is RIOT! (Version: 2024.01-devel-659-g67b381-HEAD)
2024-01-22 14:21:11,082 # Test / Hardware Details:
2024-01-22 14:21:11,085 # ========================
2024-01-22 14:21:11,085 # Cabling:
2024-01-22 14:21:11,087 # (INPUT -- OUTPUT)
2024-01-22 14:21:11,090 #   P1.14 (PB14) -- P3.0 (PD0)
2024-01-22 14:21:11,092 #   P1.15 (PB15) -- P3.9 (PD9)
2024-01-22 14:21:11,096 # Number of pull resistor values supported: 1
2024-01-22 14:21:11,099 # Number of drive strengths supported: 2
2024-01-22 14:21:11,102 # Number of slew rates supported: 1
2024-01-22 14:21:11,104 # Valid GPIO ports:
2024-01-22 14:21:11,105 # - PORT 0 (PORT A)
2024-01-22 14:21:11,107 # - PORT 1 (PORT B)
2024-01-22 14:21:11,109 # - PORT 2 (PORT C)
2024-01-22 14:21:11,110 # - PORT 3 (PORT D)
2024-01-22 14:21:11,110 # 
2024-01-22 14:21:11,113 # Testing gpio_port_pack_addr()
2024-01-22 14:21:11,116 # =============================
2024-01-22 14:21:11,116 # 
2024-01-22 14:21:11,116 # All OK
2024-01-22 14:21:11,116 # 
2024-01-22 14:21:11,118 # Testing gpip_ng_init()
2024-01-22 14:21:11,120 # ======================
2024-01-22 14:21:11,120 # 
2024-01-22 14:21:11,126 # Testing is_gpio_port_num_valid() is true for PORT_OUT and PORT_IN:
2024-01-22 14:21:11,127 # 
2024-01-22 14:21:11,130 # Testing input configurations for PIN_IN_0:
2024-01-22 14:21:11,133 # Support for input with pull up: yes
2024-01-22 14:21:11,137 # state: in, pull: up, value: on, drive: weak
2024-01-22 14:21:11,140 # Support for input with pull down: yes
2024-01-22 14:21:11,144 # state: in, pull: down, value: off, drive: weak
2024-01-22 14:21:11,148 # Support for input with pull to bus level: no
2024-01-22 14:21:11,153 # Support for floating input (no pull resistors): yes
2024-01-22 14:21:11,157 # state: in, pull: none, value: off, drive: weak
2024-01-22 14:21:11,157 # 
2024-01-22 14:21:11,161 # Testing output configurations for PIN_OUT_0:
2024-01-22 14:21:11,166 # Support for output (push-pull) with initial value of LOW: yes
2024-01-22 14:21:11,170 # state: out-pp, value: off, drive: weak
2024-01-22 14:21:11,173 # Output is indeed LOW: yes
2024-01-22 14:21:11,176 # state: out-pp, value: on, drive: weak
2024-01-22 14:21:11,180 # Output can be pushed HIGH: yes
2024-01-22 14:21:11,186 # Support for output (push-pull) with initial value of HIGH: yes
2024-01-22 14:21:11,189 # state: out-pp, value: on, drive: weak
2024-01-22 14:21:11,192 # Output is indeed HIGH: yes
2024-01-22 14:21:11,199 # Support for output (open drain with pull up) with initial value of LOW: no
2024-01-22 14:21:11,205 # Support for output (open drain with pull up) with initial value of HIGH: no
2024-01-22 14:21:11,211 # Support for output (open drain) with initial value of LOW: no
2024-01-22 14:21:11,216 # Support for output (open drain) with initial value of HIGH: no
2024-01-22 14:21:11,222 # Support for output (open source) with initial value of LOW: no
2024-01-22 14:21:11,227 # Support for output (open source) with initial value of HIGH: no
2024-01-22 14:21:11,234 # Support for output (open source with pull up) with initial value of HIGH: no
2024-01-22 14:21:11,241 # Support for output (open source with pull up) with initial value of LOW: no
2024-01-22 14:21:11,244 # Support for disconnecting GPIO: yes
2024-01-22 14:21:11,248 # state: off, pull: none, value: off, drive: weak
2024-01-22 14:21:11,252 # Output can indeed be pulled LOW: yes
2024-01-22 14:21:11,257 # Output can indeed be pulled HIGH: yes
2024-01-22 14:21:11,257 # 
2024-01-22 14:21:11,260 # Testing Reading/Writing GPIO Ports
2024-01-22 14:21:11,263 # ==================================
2024-01-22 14:21:11,263 # 
2024-01-22 14:21:11,266 # testing initial value of 0 after init
2024-01-22 14:21:11,268 # ...OK
2024-01-22 14:21:11,272 # testing setting both outputs_optional simultaneously
2024-01-22 14:21:11,274 # ...OK
2024-01-22 14:21:11,279 # testing clearing both outputs_optional simultaneously
2024-01-22 14:21:11,280 # ...OK
2024-01-22 14:21:11,284 # testing toggling first output (0 --> 1)
2024-01-22 14:21:11,285 # ...OK
2024-01-22 14:21:11,289 # testing toggling first output (1 --> 0)
2024-01-22 14:21:11,290 # ...OK
2024-01-22 14:21:11,294 # testing toggling second output (0 --> 1)
2024-01-22 14:21:11,295 # ...OK
2024-01-22 14:21:11,299 # testing toggling second output (1 --> 0)
2024-01-22 14:21:11,300 # ...OK
2024-01-22 14:21:11,306 # testing setting first output and clearing second with write
2024-01-22 14:21:11,307 # ...OK
2024-01-22 14:21:11,312 # testing setting second output and clearing first with write
2024-01-22 14:21:11,313 # ...OK
2024-01-22 14:21:11,317 # All input/output operations worked as expected
2024-01-22 14:21:11,317 # 
2024-01-22 14:21:11,319 # Testing External IRQs
2024-01-22 14:21:11,321 # =====================
2024-01-22 14:21:11,321 # 
2024-01-22 14:21:11,324 # Testing rising edge on PIN_IN_0
2024-01-22 14:21:11,327 # ... OK
2024-01-22 14:21:11,330 # Testing falling edge on PIN_IN_0
2024-01-22 14:21:11,334 # ... OK
2024-01-22 14:21:11,337 # Testing both edges on PIN_IN_0
2024-01-22 14:21:11,340 # ... OK
2024-01-22 14:21:11,345 # Testing masking of IRQs (still both edges on PIN_IN_0)
2024-01-22 14:21:11,348 # expect failed at line 661
```

File: cpu/sam0_common/periph/gpio_ll.c [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        break;
+    default:
+        return -ENOTSUP;
+    }
+
+    if (conf.drive_strength == GPIO_DRIVE_STRONG) {
+        pin_cfg |= PORT_PINCFG_DRVSTR;
+    }
+
+
```
Comment (benpicco) [inline cpu/sam0_common/periph/gpio_ll.c:129]:
```suggestion
```

File: cpu/sam0_common/periph/gpio_ll_irq.c [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+static struct isr_ctx isr_ctx[IRQS_NUMOF];
+
+extern void gpio_ll_mux(gpio_port_t port, uint8_t pin, gpio_mux_t mux);
+
+static int get_exti_num(unsigned port_num, uint8_t pin)
+{
+    if (port_num >= ARRAY_SIZE(exti_config)) {
+        return -1;
+    }
```
Comment (benpicco) [inline cpu/sam0_common/periph/gpio_ll_irq.c:76]: Can this be an `assert()`?

Comment (maribu) [reply cpu/sam0_common/periph/gpio_ll_irq.c:76]: The call sites of `get_exti_num()` actually do:

```C
    assume((unsigned)exti_num < IRQS_NUMOF);
```

That should cover this as well, but also gives the compiler information about what values are to be expected within `exti_config`.

**Comment by maribu** [comment]
> I get a different failure on `same54-xpro` though

Good catch! That would have been a nasty race condition bug that could have caused lots of pain and tears.

For me it now works correctly.

**Review by benpicco — APPROVED** [review]
Please squash

**Comment by maribu** [comment]
Thx :)

**Comment by maribu** [comment]
Improving the test in 891c627c486931683918d4ff34e606a3a4da1c71 had the side effect that lots of boards now fail the test :sweat_smile:

(But the issues it detects are indeed very real.)

**Pull request merged**